### ISSUE TITLE
Consider making root behavior more consistent

### ISSUE THREAD

**Issue by sbomer** [dotnet/linker#1541]
The SDK's `TrimmerRootAssembly` maps to `-a` which will root only the entry point for an app, or everything for a library. If the intermediate assembly is trimmable, we still want to root part of it, but `-a` would root everything in a library (see https://github.com/dotnet/sdk/issues/13999).

The `-a` behavior also means that dependencies of unused parts of the app in other assemblies could be removed, which is inconsistent with the root behavior for all other assemblies. The SDK's `TrimmerRootAssembly` should be made to consistently root the entire assembly as the name suggests.

I think the linker root behavior needs to be fixed. There should be options to:
- root the entire assembly (what -a does for an assembly without entry point)
- root the entry point (what -a does for an assembly with entry point)
- root public surface (like -r)

The last two could be combined into one option since the behavior should depend on whether it's an app or a library.

**Comment by marek-safar** [comment]
I agree this needs fixing and we could limit it to illink only if it would make things easier.

Similarly to what you are proposing, we could introduce a new option like and make the whole process less magical.

```
-root:entry|visible FILE
```

you could then further tweak what with `-p` option.
- For example `-root:visible library.dll -r copy library` would keep the file fully.
- Combinations of `-root` and `-p skip` would trigger an error message
- Similarly `-root:entry library.dll` would issue an error

**Comment by vitek-karas** [comment]
This has been implemented:
* `-a assembly library` for example roots the assembly in "library" mode. There are options for `entrypoint`, `visible`, `all`.
* This is exposed to SDK via the `RootMode` metadata on `TrimmerRootAssembly`

**Comment by MichalStrehovsky** [comment]
The targets still do this without actually rooting the whole assembly (like one would think because that's what trimmerrootassembly is supposed to do). NativeAOT needs to quirk this linker bug. I don't this this is fixed:

https://github.com/dotnet/linker/blob/b23be78205b79d60c1ea2cab5306f530a0faade2/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets#L282

**Comment by MichalStrehovsky** [comment]
> I don't think that problem is listed above (or at least I can't find it), but you're right that it's wrong. Thanks!

"The SDK's `TrimmerRootAssembly` maps to `-a` which will root only the entry point for an app, or everything for a library"

But I agree it's written a bit matter-of-factly and not as a problem. It is a problem.

**Comment by MichalStrehovsky** [comment]
I just directed someone to replace RD.XML with TrimmerRootAssembly and it didn't work because the assembly happened to be the entrypoint assembly and we don't actually root it because of this issue.

It would be good to fix this in 8. I think this became a bigger problem now that we trim everything by default and people might have actual need to root their entrypoint.
